DOCSP-40850 Java RS Write Usage Examples (#38)
mongodb/docs-java-rs: source/write-data-to-mongo.txt

.. _java-rs-write:

=====================
Write Data to MongoDB
=====================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :description: Learn how to use the {+driver-short+} to write data to MongoDB.
   :keywords: usage examples, save, crud, create, code example

.. TODO: change toctree links when other pages are merged in.
.. .. toctree::
..    :titlesonly:
..    :maxdepth: 1

..    /write/insert
..    /write/update
..    /write/replace
..    /write/delete
..    /write/bulk-write
..    /write/gridfs

Overview
--------

> On this page, you can see copyable code examples that show common
> methods to write data to MongoDB with the {+driver-short+}.

Reviewer: S: phrasing feels a little awkward. side question: what makes these methods common?

```suggestion
This page contains copyable code examples of
{+driver-short+} methods that you can use to
write data to MongoDB.
```

Author: I think they're just the main methods that people might use, but I agree that it feels a little prescriptive to call them common

.. tip::

  To learn more about any of the methods shown on this page, see the link
  provided in each section.

To use an example from this page, copy the code example into the
:ref:`sample application <java-rs-write-sample>` or your own application.
Be sure to replace all placeholders in the code examples, such as ``<connection string>``, with
the relevant values for your MongoDB deployment.

.. important:: Project Reactor Library

   This guide uses the Project Reactor library to implement ``Publisher`` and
   ``Subscriber`` methods. To learn more about the Project Reactor library and how

>    to use Reactor, see `Getting Started

Reviewer: S: is this also accurate? seems a little smoother

```suggestion
   to use it, see `Getting Started
```

.. tip::

  To learn more about any of the methods shown on this page, see the link
  provided in each section.

To use an example from this page, copy the code example into the
:ref:`sample application <java-rs-write-sample>` or your own application.
Be sure to replace all placeholders in the code examples, such as ``<connection string>``, with
the relevant values for your MongoDB deployment.

.. important:: Project Reactor Library

   This guide uses the Project Reactor library to implement ``Publisher`` and
   ``Subscriber`` methods. To learn more about the Project Reactor library and how
   to use it, see `Getting Started
   <https://projectreactor.io/docs/core/release/reference/#getting-started>`__
   in the Reactor documentation.

   In {+java-rs+}, all operations that return a ``Publisher`` are cold, which
   means that they do not execute unless you subscribe to the returned ``Publisher`` method. This guide uses the
   ``Mono.block()`` method from Reactor to subscribe to a ``Publisher`` and block the current
   thread until the ``Publisher`` reaches its terminal state. To learn more
   about {+java-rs+}, see `Reactive Streams <https://www.reactive-streams.org/>`__

>    in the {+java-rs+} documentation.

Author: Valentin and I discussed on Slack and agreed on a different wording.

   :keywords: usage examples, save, crud, create, code example

.. TODO: change toctree links when other pages are merged in.
.. .. toctree::
..    :titlesonly:
..    :maxdepth: 1

..    /write/insert
..    /write/update
..    /write/replace
..    /write/delete
..    /write/bulk-write
..    /write/gridfs

Overview
--------

This page contains copyable code examples of {+driver-short+} methods that you can use to
write data to MongoDB.

.. tip::

  To learn more about any of the methods shown on this page, see the link
  provided in each section.

To use an example from this page, copy the code example into the
:ref:`sample application <java-rs-write-sample>` or your own application.
Be sure to replace all placeholders in the code examples, such as ``<connection string>``, with
the relevant values for your MongoDB deployment.

.. important:: Project Reactor Library

   This guide uses the Project Reactor library to implement ``Publisher`` and
   ``Subscriber`` methods. To learn more about the Project Reactor library and how
   to use it, see `Getting Started
   <https://projectreactor.io/docs/core/release/reference/#getting-started>`__
   in the Reactor documentation.

> In {+java-rs+}, all operations that return a ``Publisher`` do not execute

Reviewer 2: I see "In ​, all operations" in staging. Something is wrong with `{+java-rs+}` here and in other places it is used at.

Reviewer 2: ✅ Done.

> unless you subscribe to the returned ``Publisher`` method. This guide uses the
> ``block()`` method to subscribe to all ``Publisher`` methods. To learn more

Reviewer 2: One can subscribe to a `Publisher`, not to a method (regardless of whether it's a method on `Publisher` or on any other type).

```suggestion
unless you subscribe to the returned ``Publisher``. This guide uses the
``Mono.block()`` method to subscribe to a ``Publisher`` and block until it reaches its terminal state. To learn more
```

Reviewer 2: > and block the ``Publisher`` until it reaches its terminal state

`block` does not block a `Publisher`, it blocks the current thread. You can either say what I proposed "and block until it reaches its terminal state", or "and block the current thread until the `Publisher` reaches its terminal state".

Reviewer 2: - The Java drivers team also discussed and decided that a note like this should be moved into the `IMPORTANT` section that you already have on the page.
- Additionally, it may be worth mentioning the [cold `Publisher`](https://projectreactor.io/docs/core/release/reference/#reactor.hotCold) term there. We can say that all `Publishers` returned from the driver API are cold, and then explain that it means operations do not execute unless you subscribe bla-bla-bla.

Reviewer 2: > all operations that return a Publisher are cold

It's the `Publisher`s that are cold, not operations. I'll suggest an edit for the whole `IMPORTANT` note as a new comment, as I think, it's also worth changing the wording in the first part of the note, where it talks about the Project Reactor.

.. TODO: change toctree links when other pages are merged in.
.. .. toctree::
..    :titlesonly:
..    :maxdepth: 1

..    /write/insert
..    /write/update
..    /write/replace
..    /write/delete
..    /write/bulk-write
..    /write/gridfs

Overview
--------

This page contains copyable code examples of {+driver-short+} methods that you can use to
write data to MongoDB.

.. tip::

  To learn more about any of the methods shown on this page, see the link
  provided in each section.

To use an example from this page, copy the code example into the
:ref:`sample application <java-rs-write-sample>` or your own application.
Be sure to replace all placeholders in the code examples, such as ``<connection string>``, with
the relevant values for your MongoDB deployment.

This guide uses the Project Reactor library to consume ``Publishers`` returned
by the {+driver-short+} methods. To learn more about the Project Reactor library
and how to use it, see `Getting Started <https://projectreactor.io/docs/core/release/reference/#getting-started>`__
in the Reactor documentation. 

There are also other ways to consume ``Publishers``. You can use one of many alternative libraries such as

> RxJava <https://github.com/ReactiveX/RxJava>__ to consume ``Publishers``, or consume ``Publishers``

Reviewer 2: This hyperlink appears broken:
<img width="479" alt="Screenshot 2024-07-18 at 10 16 03" src="https://github.com/user-attachments/assets/bc3372a9-40c8-4a28-bbd5-d2849a0a2161">
